### PR TITLE
Feat/paych cancel fails with valid conditions 2678

### LINKED ISSUE #2678: Payment Channel Cancel Fails if Condition exists and is valid
Blocked on #2677

### Description

We don't want payers to be able to harass targets by continually canceling a channel forcing the target to respond with more expensive redeem calls. As of #2677, we will be storing whether the target has successfully redeemed against a payment channel and the last condition they've used. We can use this information to block frivolous calls to cancel.

The `paymentbroker.Actor.Cancel` method should fail if:

1. The payment channel `redeemed` flag is true, and there is no condition.
2. The payment channel `redeemed` flag is true, a condition exists, and no revert error is returned when the condition is converted into a message and called (i.e. the condition is still valid)

### Acceptance criteria

1. The payment channel respect the conditions described above.
2. Integration tests exist for both code paths.

### Risks + pitfalls

Sending a message and continuing only when an error is return, is unnatural and a little risky. Some thought care be given to which conditions constitute a validation failure versus a general system failure. This should be the difference between a revert error and a fault, but there might be scenarios that don't fit that categorization.

### PR DESCRIPTION
**Pull request by rosalinekarr**
# Problem

The current implementation of payment channel cancel allows payers to harass payees by repeatedly cancelling channels and forcing payees to reopen them by redeeming.

# Solution

Only allow payers to cancel payment channels when the channel either hasn't been redeemed yet or the last redeem included invalid conditions.

Resolves #2678

### CODE REVIEW

File: actor/builtin/paymentbroker/paymentbroker_test.go [hunk 1]
```diff
@@ -825,6 +825,60 @@ func TestPaymentBrokerCancel(t *testing.T) {
...
+	toAddress := addrGetter()
+	method := "paramsNotZero"
+	addrParam := addrGetter()
+	sectorIdParam := uint64(6)
+	payerParams := []interface{}{addrParam, sectorIdParam}
+	blockHeightParam := types.NewBlockHeight(43)
+	redeemerParams := []interface{}{blockHeightParam}
+
+	sys := setup(t)
+	require.NoError(t, sys.st.SetActor(context.TODO(), toAddress, actor.NewActor(pbTestActorCid, types.NewZeroAttoFIL())))
```
Comment (frrist) [inline actor/builtin/paymentbroker/paymentbroker_test.go]: baby nit: call this `context.Background()` unless we intend to alter this context later. Code should use context.TODO when it's unclear which Context to use or it is not yet available. Ditto this below.

File: actor/builtin/paymentbroker/paymentbroker_test.go [hunk 2]
```diff
@@ -825,6 +825,60 @@ func TestPaymentBrokerCancel(t *testing.T) {
...
+
+	// Redeem the payment channel with bad params and expect invalid condition error
+	condition := &types.Predicate{To: toAddress, Method: method, Params: payerParams}
+	result, err := sys.applySignatureMessage(sys.target, 200, types.NewBlockHeight(0), 0, "redeem", 0, condition)
+	require.NoError(t, err)
+	require.Error(t, result.ExecutionError)
+	require.EqualValues(t, errors.CodeError(result.ExecutionError), ErrConditionInvalid)
```
Comment (frrist) [inline actor/builtin/paymentbroker/paymentbroker_test.go]: The params are backwards:
```suggestion
require.EqualValues(t, ErrConditionInvalid, errors.CodeError(result.ExecutionError))
```

File: actor/builtin/paymentbroker/paymentbroker.go [hunk 3]
```diff
@@ -414,6 +414,20 @@ func (pb *Actor) Cancel(vmctx exec.VMContext, chid *types.ChannelID) (uint8, err
 			return errors.NewFaultError("Expected PaymentChannel from channels lookup")
 		}
 
+		if channel.Redeemed {
+			if channel.Condition == nil {
+				return errors.NewFaultError("Channel cannot be cancelled due to successful redeem")
+			} else {
+				err := checkCondition(vmctx, channel, channel.Condition, []interface{}{})
+				if err != nil && !errors.IsFault(err) {
```
Comment (acruikshank) [inline actor/builtin/paymentbroker/paymentbroker.go]: BLOCKING: If the error is not a revert error (`IsRevert`) it should be returned. Faults (and bad programmer) errors should be returned immediately.

File: actor/builtin/paymentbroker/paymentbroker.go [hunk 4]
```diff
@@ -414,6 +414,20 @@ func (pb *Actor) Cancel(vmctx exec.VMContext, chid *types.ChannelID) (uint8, err
...
+					return err
+				}
+				if err == nil {
+					return errors.NewFaultError("Channel cannot be cancelled due to successful redeem")
```
Comment (acruikshank) [inline actor/builtin/paymentbroker/paymentbroker.go]: BLOCKING: a passing condition here should produce a coded revert error. Faults indicate that some aspect of internal state is corrupt and block processing should be abandoned altogether.

File: actor/builtin/paymentbroker/paymentbroker.go [hunk 5]
```diff
@@ -414,6 +414,20 @@ func (pb *Actor) Cancel(vmctx exec.VMContext, chid *types.ChannelID) (uint8, err
 			return errors.NewFaultError("Expected PaymentChannel from channels lookup")
 		}
 
+		if channel.Redeemed {
+			if channel.Condition == nil {
+				return errors.NewFaultError("Channel cannot be cancelled due to successful redeem")
```
Comment (acruikshank) [inline actor/builtin/paymentbroker/paymentbroker.go]: BLOCKING: This error must be a revert error and should be coded.

File: actor/builtin/paymentbroker/paymentbroker_test.go [hunk 6]
```diff
@@ -825,6 +825,60 @@ func TestPaymentBrokerCancel(t *testing.T) {
...
+	payerParams := []interface{}{toAddress, sectorIdParam}
+
+	sys := setup(t)
+	require.NoError(t, sys.st.SetActor(context.Background(), toAddress, actor.NewActor(pbTestActorCid, types.NewZeroAttoFIL())))
+
+	// Redeem the payment channel with bad params and expect invalid condition error
+	condition := &types.Predicate{To: toAddress, Method: method, Params: payerParams}
+	result, err := sys.applySignatureMessage(sys.target, 200, types.NewBlockHeight(0), 0, "redeem", 0, condition)
+	require.NoError(t, err)
+	require.Error(t, result.ExecutionError)
```
Comment (acruikshank) [inline actor/builtin/paymentbroker/paymentbroker_test.go]: BLOCKING: The cancel succeeds below because the redeem fails and the `Redeemed` flag never gets set. In order to demonstrate the cancel can succeed after a condition is set:

1. Redeem must be called with a condition successfully.
2. Establish that the condition and redeem flag are set in the payment channel.
3. Change internal state occur to invalidate the condition.
4. Call Redeem and establish it succeeds.

The easiest way to alter internal state is add a method to PBTestActor that succeeds or fails based on some flag. Ideally this would be in actor state, but it's probably ok to create a field on PBTestActor itself.

File: actor/builtin/paymentbroker/paymentbroker_test.go [hunk 7]
```diff
@@ -825,6 +825,60 @@ func TestPaymentBrokerCancel(t *testing.T) {
...
+	require.EqualValues(t, ErrConditionInvalid, errors.CodeError(result.ExecutionError))
+
+	// Attempt to Cancel and expects success
+	pdata := core.MustConvertParams(sys.channelID)
+	msg := types.NewMessage(sys.payer, address.PaymentBrokerAddress, 1, types.NewAttoFILFromFIL(1000), "cancel", pdata)
+	_, err = sys.ApplyMessage(msg, 100)
+	assert.NoError(t, err)
+}
```
Comment (acruikshank) [inline actor/builtin/paymentbroker/paymentbroker_test.go]: BLOCKING: We need a test that tests that cancel fails when the payment channel is redeemed with a nil condition.

File: actor/builtin/paymentbroker/paymentbroker_test.go [hunk 8]
```diff
@@ -825,6 +825,60 @@ func TestPaymentBrokerCancel(t *testing.T) {
...
+	condition := &types.Predicate{To: toAddress, Method: method, Params: payerParams}
+	result, err := sys.applySignatureMessage(sys.target, 100, types.NewBlockHeight(0), 0, "redeem", 0, condition, redeemerParams...)
+	require.NoError(t, err)
+	require.NoError(t, result.ExecutionError)
+
+	// Attempts to Cancel and expects failure
+	pdata := core.MustConvertParams(sys.channelID)
+	msg := types.NewMessage(sys.payer, address.PaymentBrokerAddress, 1, types.NewAttoFILFromFIL(1000), "cancel", pdata)
+	_, err = sys.ApplyMessage(msg, 100)
+	assert.Error(t, err)
```
Comment (acruikshank) [inline actor/builtin/paymentbroker/paymentbroker_test.go]: BLOCKING: Please assert the error code.

File: actor/builtin/paymentbroker/paymentbroker.go (hunk 3 above)
Comment (acruikshank) [reply actor/builtin/paymentbroker/paymentbroker.go]: Specifically
1. Check err == nil. if so return coded revert error.
2. Check if `!errors.IsRevert(err)`. If so return error unmodified.,
3. else continue.

File: actor/builtin/paymentbroker/paymentbroker.go [hunk 9]
```diff
@@ -414,6 +414,20 @@ func (pb *Actor) Cancel(vmctx exec.VMContext, chid *types.ChannelID) (uint8, err
...
+			} else {
+				err := checkCondition(vmctx, channel, channel.Condition, []interface{}{})
```
Comment (acruikshank) [inline actor/builtin/paymentbroker/paymentbroker.go]: This is a little confusing. It would be better to extract the part of `checkCondtion` that makes the call from the part that handles storing the condition, and just call that here.

File: actor/builtin/paymentbroker/paymentbroker.go [hunk 10]
```diff
@@ -414,6 +414,20 @@ func (pb *Actor) Cancel(vmctx exec.VMContext, chid *types.ChannelID) (uint8, err
 			return errors.NewFaultError("Expected PaymentChannel from channels lookup")
 		}
 
+		if channel.Redeemed {
```
Comment (acruikshank) [inline actor/builtin/paymentbroker/paymentbroker.go]: Please explain this logic in the method comments.

File: actor/builtin/paymentbroker/paymentbroker.go (hunk 3 above)
Comment (rosalinekarr) [reply actor/builtin/paymentbroker/paymentbroker.go]: `IsRevert` isn't a method implemented anywhere. Do you want me to write it?

Comment (rosalinekarr) [reply actor/builtin/paymentbroker/paymentbroker.go]: Do you mean `ShouldRevert`?

Comment (rosalinekarr) [reply actor/builtin/paymentbroker/paymentbroker.go]: (Answered offline: Yes)
